config/initializers/000-development_reload_warnings.rb

```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+# Development helper which prints a warning when you edit a non-autoloaded ruby file.
+# These include initializers, middleware, plugin.rb files, and more.
+# Launch the server with AUTO_RESTART=1 to automate restarts.
+if Rails.env.development? && !Rails.configuration.cache_classes
+  paths = [
+    *Dir["#{Rails.root}/app/*"].reject { |path| path.end_with? "/assets" },
+    "#{Rails.root}/config",
+    "#{Rails.root}/lib",
+    "#{Rails.root}/plugins"
+  ]
+
+  auto_restart = ENV["AUTO_RESTART"]
+
+  Listen.to(*paths, only: /\.rb$/) do |modified, added, removed|
```

Comment:
Are we able to listen to events only when the contents of the file have changed? Right now, the warning is being triggered even when I save without making any changes.

Reply:
Unfortunately not easily - the `Listen` gem notifies us about changes (via various OS-specific hooks), and then gives us the filenames. 'changes' in this case includes changes to the file metadata (e.g. modified time). Technically I guess we could load every file into memory and check for diffs, but I think it would make things a lot slower.

This method of 'checking modified time' is the same thing which Rails uses to trigger code reloads.